src/annotator/anchoring/range.js

```diff
@@ -0,0 +1,494 @@
+      ) {
+        normalRange.start.splitText(range.endOffset - range.startOffset);
+      }
+      normalRange.end = normalRange.start;
+    } else {
+      // No, the end of the selection is in a separate text element
+      // does the end need to be cut?
+      if (range.end.nodeValue.length > range.endOffset) {
+        range.end.splitText(range.endOffset);
+      }
+      normalRange.end = range.end;
+    }
+
+    // Make sure the common ancestor is an element node.
+    normalRange.commonAncestor = this.commonAncestorContainer;
+    while (normalRange.commonAncestor.nodeType !== Node.ELEMENT_NODE) {
+      normalRange.commonAncestor = normalRange.commonAncestor.parentNode;
+    }
+
+    return new Range.NormalizedRange(normalRange);
+  }
+
+  /**
+   * Creates a range suitable for storage.
+   *
+   * root           - A root Element from which to anchor the serialization.
+   * ignoreSelector - A selector String of elements to ignore. For example
+   *                  elements injected by the annotator.
+   *
+   * Returns an instance of SerializedRange.
+   */
+  serialize(root, ignoreSelector) {
+    return this.normalize().serialize(root, ignoreSelector);
+  }
+};
+
+/**
+ * A normalized range is most commonly used throughout the annotator.
+ * its the result of a deserialized SerializedRange or a BrowserRange with
+ * out browser inconsistencies.
+ */
+Range.NormalizedRange = class NormalizedRange {
+  /**
+   * Creates an instance of a NormalizedRange.
+   *
+   * This is usually created by calling the .normalize() method on one of the
+   * other Range classes rather than manually.
+   *
+   * obj - An Object literal. Should have the following properties.
+   *       commonAncestor: A Element that encompasses both the start and end nodes
+   *       start:          The first TextNode in the range.
+   *       end             The last TextNode in the range.
+   *
+   * Returns an instance of NormalizedRange.
+   */
+  constructor(obj) {
+    this.commonAncestor = obj.commonAncestor;
+    this.start = obj.start;
+    this.end = obj.end;
+  }
+
+  /**
+   * For API consistency.
+   *
+   * Returns itself.
+   */
+  normalize() {
+    return this;
+  }
+
+  /**
+   * Limits the nodes within the NormalizedRange to those contained
+   * withing the bounds parameter. It returns an updated range with all
+   * properties updated. NOTE: Method returns null if all nodes fall outside
+   * of the bounds.
+   *
+   * bounds - An Element to limit the range to.
+   *
+   * Returns updated self or null.
+   */
+  limit(bounds) {
+    const nodes = $.grep(
+      this.textNodes(),
+      node => node.parentNode === bounds || $.contains(bounds, node.parentNode)
+    );
+    if (!nodes.length) {
+      return null;
+    }
+
+    this.start = nodes[0];
+    this.end = nodes[nodes.length - 1];
+
+    const startParents = $(this.start).parents();
+
+    for (let parent of Array.from($(this.end).parents())) {
```

Comment:
Can you check whether each of the `Array.from(...)` calls in this file are needed? If you used `decaffeinate` to produce this file it will have generated them quite liberally, including in cases where we know the value is definitely going to be an array.